pilot/pkg/serviceregistry/external/conversion.go

```diff
@@ -15,99 +15,107 @@
+	for _, port := range externalService.Ports {
+		svcPorts = append(svcPorts, convertPort(port))
+	}
 
-		svcPorts := make(model.PortList, 0, len(externalService.Ports))
-		for _, port := range externalService.Ports {
-			svcPorts = append(svcPorts, convertPort(port))
+	for _, host := range externalService.Hosts {
+		// set address if host is an IP or CIDR prefix
+		var address string
+		if _, _, cidrErr := net.ParseCIDR(host); cidrErr == nil || net.ParseIP(host) != nil {
+			address = host
+			// FIXME: create common function for CIDR prefix to metrics friendly name?
+			host = strings.Replace(host, "/", "_", -1) // make hostname easy to parse for metrics
 		}
 
 		out = append(out, &model.Service{
 			MeshExternal: true,
 			Hostname:     host,
-			// TODO: add Address if the host is a CIDR
-			Ports:      svcPorts,
-			Resolution: resolution,
+			Address:      address,
+			Ports:        svcPorts,
+			Resolution:   resolution,
 		})
 	}
 
 	return out
 }
 
-func convertNetworkEndpoint(services []*model.Service, servicePort *networking.Port, endpoint *networking.ExternalService_Endpoint) []*model.ServiceInstance {
-	out := make([]*model.ServiceInstance, 0)
-	for _, service := range services {
-		instancePort := endpoint.Ports[servicePort.Name]
-		if instancePort == 0 {
-			instancePort = servicePort.Number
-		}
+func convertEndpoint(service *model.Service, servicePort *networking.Port,
+	endpoint *networking.ExternalService_Endpoint) *model.ServiceInstance {
 
-		serviceInstance := &model.ServiceInstance{
-			Endpoint: model.NetworkEndpoint{
-				Address:     endpoint.Address,
-				Port:        int(instancePort),
-				ServicePort: convertPort(servicePort),
-			},
-			// TODO AvailabilityZone
-			Service: service,
-			Labels:  endpoint.Labels,
-		}
-		out = append(out, serviceInstance)
+	instancePort := endpoint.Ports[servicePort.Name]
+	if instancePort == 0 {
+		instancePort = servicePort.Number
+	}
+
+	return &model.ServiceInstance{
+		Endpoint: model.NetworkEndpoint{
+			Address:     endpoint.Address,
+			Port:        int(instancePort),
+			ServicePort: convertPort(servicePort),
+		},
+		// TODO AvailabilityZone, ServiceAccount
+		Service: service,
+		Labels:  endpoint.Labels,
 	}
-	return out
 }
 
 func convertInstances(externalService *networking.ExternalService) []*model.ServiceInstance {
 	out := make([]*model.ServiceInstance, 0)
-	services := convertService(externalService)
-
-	for _, endpoint := range externalService.Endpoints {
+	for _, service := range convertServices(externalService) {
 		for _, servicePort := range externalService.Ports {
-
-			out = append(out, convertNetworkEndpoint(services, servicePort, endpoint)...)
+			if len(externalService.Endpoints) == 0 &&
+				externalService.Discovery == networking.ExternalService_DNS {
+				// when external service has discovery type DNS and no endpoints
+				// we create endpoints from external service hosts field
+				for _, host := range externalService.Hosts {
```

Comment:
@ZackButcher @rshriram: this should fix the external service DNS + no endpoints issue you were seeing.

Reply:
FYI @diemtvu